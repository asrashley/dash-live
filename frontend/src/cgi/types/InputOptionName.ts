export interface InputOptionName {
    shortName: string;
    fullName: string;
    cgiName: string;
}