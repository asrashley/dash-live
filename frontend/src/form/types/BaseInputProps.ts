import type { ReadonlySignal } from "@preact/signals";
import { StaticInputProps } from "./StaticInputProps";

export type BaseInputProps = {
  name: string;
  id: string;
  type: StaticInputProps["type"];
  className: string | ReadonlySignal<string>;
  title: string;
  placeholder?: string;
  "aria-describedby": string;
  disabled: ReadonlySignal<boolean>;
  onInput: (ev: Event) => void;
};

