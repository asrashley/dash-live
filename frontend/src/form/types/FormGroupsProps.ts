import { InputFormGroup } from "./InputFormGroup";
import { FormRowMode } from "./FormRowMode";
import { SetValueFunc } from "./SetValueFunc";
import { InputProps } from "./InputProps";

export interface FormGroupsProps extends Pick<InputProps, "data" | "disabledFields"> {
  groups: InputFormGroup[];
  expand?: string;
  mode: FormRowMode;
  layout?: number[];
  setValue: SetValueFunc;
}
