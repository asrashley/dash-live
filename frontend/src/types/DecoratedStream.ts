import { Stream } from "./Stream";
import { StreamTrack } from "./StreamTrack";


export interface DecoratedStream extends Stream {
    tracks: StreamTrack[];
}
