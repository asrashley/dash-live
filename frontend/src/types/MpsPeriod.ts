import { MpsTrack } from './MpsTrack';

export type MpsPeriod = {
    duration: string;
    ordering: number;
    parent: number;
    pid: string;
    pk: number | string;
    start: string;
    stream: number;
    tracks: MpsTrack[];
    new?: boolean;
};
