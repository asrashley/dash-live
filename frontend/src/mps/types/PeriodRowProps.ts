import { MpsPeriod } from "../../types/MpsPeriod";

export interface PeriodRowProps {
  className?: string;
  index: number;
  item: MpsPeriod;
}
