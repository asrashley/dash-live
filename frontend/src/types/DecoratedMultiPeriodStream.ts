import { MpsPeriod } from "./MpsPeriod";
import { MultiPeriodStream } from "./MultiPeriodStream";


export type DecoratedMultiPeriodStream = Omit<MultiPeriodStream, 'periods'> & {
    periods: MpsPeriod[];
    modified: boolean;
    lastModified: number;
};
