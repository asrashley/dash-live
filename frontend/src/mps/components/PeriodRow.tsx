import { type JSX } from "preact";
import { useComputed } from "@preact/signals";
import { useContext, useCallback } from "preact/hooks";

import { AppStateContext } from "../../appState";
import { TextInput } from "../../components/TextInput";
import { TimeDeltaInput } from "../../components/TimeDeltaInput";
import { AllStreamsContext } from "../../hooks/useAllStreams";
import { MultiPeriodModelContext, MpsPeriodValidationErrors, UseMultiPeriodStreamHook } from "../../hooks/useMultiPeriodStream";
import { PeriodRowProps } from "../types/PeriodRowProps";
import { PeriodOrder } from "./PeriodOrder";
import { rowColours } from "./rowColours";
import { StreamSelection } from "./StreamSelection";
import { TrackSelectionButton } from "./TrackSelectionButton";
import { DecoratedStream } from "../../types/DecoratedStream";
import { MpsTrack } from "../../types/MpsTrack";

function setPeriodStream(periodPk: string | number, stream: DecoratedStream, currentTracks: MpsTrack[], modifyPeriod: UseMultiPeriodStreamHook["modifyPeriod"]) {
  let tracks: MpsTrack[] = [...currentTracks];
  const tids = new Set(tracks.map((tk) => tk.track_id));
  const activeContent = new Set(
    tracks.filter((tk) => tk.enabled).map((tk) => tk.content_type)
  );
  const streamTracks = new Set<number>();
  stream.tracks.forEach((tk) => {
    streamTracks.add(tk.track_id);
    if (!tids.has(tk.track_id)) {
      const newTk: MpsTrack = {
        lang: null,
        ...tk,
        role: activeContent.has(tk.content_type) ? "alternate": "main",
        enabled: !activeContent.has(tk.content_type),
        encrypted: false,
      };
      if (newTk.enabled) {
        activeContent.add(newTk.content_type);
      }
      tracks.push(newTk);
    }
  });
  tracks = tracks.filter(tk => streamTracks.has(tk.track_id));
  tracks.sort((a, b) => a.track_id - b.track_id);

  modifyPeriod({
    periodPk,
    period: {
      stream: stream.pk,
      duration: stream.duration,
    },
    tracks,
  });
}

export function PeriodRow({ className = "", index, item: period }: PeriodRowProps) {
  const { streamsMap } = useContext(AllStreamsContext);
  const { dialog } = useContext(AppStateContext);
  const { errors, modifyPeriod, addPeriod, removePeriod } = useContext(
    MultiPeriodModelContext
  );
  const prdErrors = useComputed<MpsPeriodValidationErrors>(() => errors?.[period?.pid] ?? {});
  const currentStream = streamsMap.value.get(`${period.stream}`);
  const { pid, pk, start, duration } = period;

  const selectTracks = useCallback(() => {
    if (currentStream === undefined) {
      return;
    }
    dialog.value = {
      backdrop: true,
      trackPicker: {
        pk,
        pid,
        guest: false,
        stream: currentStream,
      },
    };
  }, [dialog, pid, pk, currentStream]);

  const setPid = useCallback(
    (ev: JSX.TargetedEvent<HTMLInputElement>) => {
      modifyPeriod({
        periodPk: pk,
        period: {
          pid: (ev.target as HTMLInputElement).value,
        },
      });
    },
    [modifyPeriod, pk]
  );

  const setField = useCallback(
    (name: string, value: string) => {
      const fieldName = name.split("_")[0];
      modifyPeriod({
        periodPk: pk,
        period: {
          [fieldName]: value,
        },
      });
    },
    [modifyPeriod, pk]
  );

  const setStream = useCallback(
    ({ value }: {name: string, value: number}) => {
      const stream = streamsMap.value.get(`${value}`);
      setPeriodStream(pk, stream, period.tracks, modifyPeriod);
    },
    [modifyPeriod, period, pk, streamsMap]
  );

  const deletePeriodBtn = useCallback(() => {
    removePeriod(pk);
  }, [pk, removePeriod]);

  const clsNames = `row mt-1 p-1 ${rowColours[index % rowColours.length]} ${className}`;

  return <li className={clsNames}>
    <div className="col period-ordering">
      <PeriodOrder addPeriod={addPeriod} deletePeriod={deletePeriodBtn} />
    </div>
    <div className="col period-id">
      <TextInput
        value={pid}
        name={`pid_${pk}`}
        onInput={setPid}
        error={prdErrors.value.pid}
        required />
    </div>
    <div class="col period-stream">
      <StreamSelection
        name={`stream_${pk}`}
        value={currentStream}
        onChange={setStream}
        error={prdErrors.value.stream}
        required />
    </div>
    <div class="col period-start">
      <TimeDeltaInput
        value={start}
        name={`start_${pk}`}
        onChange={setField}
        error={prdErrors.value.start}
        required />
    </div>
    <div class="col period-duration">
      <TimeDeltaInput
        value={duration}
        name={`duration_${pk}`}
        onChange={setField}
        min="00:00:01"
        error={prdErrors.value.duration}
        required />
    </div>
    <TrackSelectionButton period={period} stream={currentStream} selectTracks={selectTracks} />
  </li>;
}
