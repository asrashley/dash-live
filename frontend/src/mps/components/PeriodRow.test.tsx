import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import { fireEvent } from "@testing-library/preact";
import { mock } from "vitest-mock-extended";
import userEvent from "@testing-library/user-event";
import { signal } from "@preact/signals";

import { renderWithProviders } from "../../test/renderWithProviders";
import { MpsPeriod } from "../../types/MpsPeriod";
import {
  AllStreamsContext,
  decorateAllStreams,
  UseAllStreamsHook,
} from "../../hooks/useAllStreams";
import { DecoratedStream } from "../../types/DecoratedStream";
import { PeriodRow } from "./PeriodRow";

import { streams } from "../../test/fixtures/streams.json";
import { model } from "../../test/fixtures/multi-period-streams/demo.json";
import {
  MpsModelValidationErrors,
  MultiPeriodModelContext,
  UseMultiPeriodStreamHook,
} from "../../hooks/useMultiPeriodStream";
import { DecoratedMultiPeriodStream } from "../../types/DecoratedMultiPeriodStream";

describe("PeriodRow component", () => {
  const allStreams = signal<DecoratedStream[]>([]);
  const loaded = signal<boolean>(false);
  const streamsMap = signal<Map<string, DecoratedStream>>(new Map());
  const error = signal<string | null>(null);
  const allStreamsHook: UseAllStreamsHook = {
    allStreams,
    loaded,
    streamsMap,
    error,
  };
  let multiPeriodStreamHook: UseMultiPeriodStreamHook;

  beforeEach(() => {
    loaded.value = true;
    error.value = null;
    allStreams.value = decorateAllStreams(streams);
    const sMap = new Map<string, DecoratedStream>();
    allStreams.value.forEach((item) => {
      sMap.set(`${item.pk}`, item);
    });
    streamsMap.value = sMap;
    multiPeriodStreamHook = mock<UseMultiPeriodStreamHook>({
      loaded: signal<string | undefined>(),
      model: signal<DecoratedMultiPeriodStream>(),
      modified: signal<boolean>(false),
      errors: signal<MpsModelValidationErrors>({}),
      isValid: signal<boolean>(false),
    });
    });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test.each(model.periods)(
    "matches snapshot for period $pid",
    async (item: MpsPeriod) => {
      const { asFragment, findByText } = renderWithProviders(
        <AllStreamsContext.Provider value={allStreamsHook}>
          <MultiPeriodModelContext.Provider value={multiPeriodStreamHook}>
            <PeriodRow index={2} item={item} className="guest-period-row" />
          </MultiPeriodModelContext.Provider>
        </AllStreamsContext.Provider>
      );
      const decStream = streamsMap.value.get(`${item.stream}`);
      expect(decStream).toBeDefined();
      await findByText(decStream.title);
      expect(asFragment()).toMatchSnapshot();
    }
  );

  test("opens track view dialog", () => {
    const period = model.periods[0];
    const { getBySelector, appState: state } = renderWithProviders(
      <AllStreamsContext.Provider value={allStreamsHook}>
        <MultiPeriodModelContext.Provider value={multiPeriodStreamHook}>
          <PeriodRow index={2} item={period} className="guest-period-row" />
        </MultiPeriodModelContext.Provider>
      </AllStreamsContext.Provider>
    );
    const elt = getBySelector(".period-tracks > .btn") as HTMLButtonElement;
    fireEvent.click(elt);
    const decStream = streamsMap.value.get(`${period.stream}`);
    expect(decStream).toBeDefined();
    expect(state.dialog.value).toEqual({
      backdrop: true,
      trackPicker: {
        pk: period.pk,
        pid: period.pid,
        guest: false,
        stream: decStream,
      },
    });
  });

  test('can set period ID', async () => {
    const user = userEvent.setup();
    const period = model.periods[0];
    const { getBySelector } = renderWithProviders(
      <AllStreamsContext.Provider value={allStreamsHook}>
        <MultiPeriodModelContext.Provider value={multiPeriodStreamHook}>
          <PeriodRow index={2} item={period} className="guest-period-row" />
        </MultiPeriodModelContext.Provider>
      </AllStreamsContext.Provider>
    );
    const pidElt = getBySelector(".period-id input") as HTMLInputElement;
    await user.clear(pidElt);
    await user.click(pidElt);
    await user.type(pidElt, 'hello{enter}');
    expect(multiPeriodStreamHook.modifyPeriod).toHaveBeenLastCalledWith({
      periodPk: period.pk,
      period: {
        pid: 'hello',
      },
    });
  });

  test('can select stream', async () => {
    const user = userEvent.setup();
    const period = model.periods[0];
    const { getBySelector } = renderWithProviders(
      <AllStreamsContext.Provider value={allStreamsHook}>
        <MultiPeriodModelContext.Provider value={multiPeriodStreamHook}>
          <PeriodRow index={2} item={period} className="guest-period-row" />
        </MultiPeriodModelContext.Provider>
      </AllStreamsContext.Provider>
    );
    const selElt = getBySelector(".period-stream select") as HTMLSelectElement;
    expect(selElt.value).toEqual(`${period.stream}`);
    const newPk = (period.stream + 1) % allStreams.value.length;
    const newStream = streamsMap.value.get(`${newPk}`);
    expect(newStream).toBeDefined();
    await user.selectOptions(selElt, [newStream.title]);
    expect(multiPeriodStreamHook.modifyPeriod).toHaveBeenLastCalledWith({
      periodPk: period.pk,
      period: {
        stream: newPk,
        duration: expect.any(String),
      },
      tracks: expect.any(Array),
    });
  });
});
