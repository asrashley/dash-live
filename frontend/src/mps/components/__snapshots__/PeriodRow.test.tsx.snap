// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PeriodRow component > matches snapshot for period 'p1' 1`] = `
<DocumentFragment>
  <li
    class="row mt-1 p-1 bg-secondary-subtle guest-period-row"
  >
    <div
      class="col period-ordering"
    >
      <div
        class="dropdown"
      >
        <a
          aria-expanded="false"
          class=" dropdown-toggle"
          href="#"
          role="button"
        >
          <span
            aria-label="three-dots icon"
            class="bi icon bi-three-dots "
            role="img"
          />
        </a>
        <ul
          class="dropdown-menu "
        >
          <li
            data-testid="ddi_0"
          >
            <a
              class="dropdown-item"
              href="#"
            >
              Add another Period
            </a>
          </li>
          <li
            data-testid="ddi_1"
          >
            <a
              class="dropdown-item"
              href="#"
            >
              Delete Period
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="col period-id"
    >
      <input
        class="form-control is-valid"
        id="field-pid_1"
        name="pid_1"
        required=""
        type="text"
      />
    </div>
    <div
      class="col period-stream"
    >
      <select
        class="form-select is-valid"
        name="stream_1"
        required=""
      >
        <option
          value=""
        >
          --Select a stream--
        </option>
        <option
          value="1"
        >
          Big Buck Bunny
        </option>
        <option
          value="2"
        >
          Tears of Steel
        </option>
      </select>
    </div>
    <div
      class="col period-start"
    >
      <input
        class="form-control is-valid"
        min="00:00:00"
        name="start_1"
        step="1"
        type="time"
      />
    </div>
    <div
      class="col period-duration"
    >
      <input
        class="form-control is-valid"
        min="00:00:01"
        name="duration_1"
        step="1"
        type="time"
      />
    </div>
    <div
      class="col period-tracks"
    >
      <button
        aria-disabled="false"
        class="btn btn-sm m-1 btn-warning"
      >
        0/3 tracks
      </button>
    </div>
  </li>
</DocumentFragment>
`;

exports[`PeriodRow component > matches snapshot for period 'p2' 1`] = `
<DocumentFragment>
  <li
    class="row mt-1 p-1 bg-secondary-subtle guest-period-row"
  >
    <div
      class="col period-ordering"
    >
      <div
        class="dropdown"
      >
        <a
          aria-expanded="false"
          class=" dropdown-toggle"
          href="#"
          role="button"
        >
          <span
            aria-label="three-dots icon"
            class="bi icon bi-three-dots "
            role="img"
          />
        </a>
        <ul
          class="dropdown-menu "
        >
          <li
            data-testid="ddi_0"
          >
            <a
              class="dropdown-item"
              href="#"
            >
              Add another Period
            </a>
          </li>
          <li
            data-testid="ddi_1"
          >
            <a
              class="dropdown-item"
              href="#"
            >
              Delete Period
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="col period-id"
    >
      <input
        class="form-control is-valid"
        id="field-pid_2"
        name="pid_2"
        required=""
        type="text"
      />
    </div>
    <div
      class="col period-stream"
    >
      <select
        class="form-select is-valid"
        name="stream_2"
        required=""
      >
        <option
          value=""
        >
          --Select a stream--
        </option>
        <option
          value="1"
        >
          Big Buck Bunny
        </option>
        <option
          value="2"
        >
          Tears of Steel
        </option>
      </select>
    </div>
    <div
      class="col period-start"
    >
      <input
        class="form-control is-valid"
        min="00:00:00"
        name="start_2"
        step="1"
        type="time"
      />
    </div>
    <div
      class="col period-duration"
    >
      <input
        class="form-control is-valid"
        min="00:00:01"
        name="duration_2"
        step="1"
        type="time"
      />
    </div>
    <div
      class="col period-tracks"
    >
      <button
        aria-disabled="false"
        class="btn btn-sm m-1 btn-warning"
      >
        0/3 tracks
      </button>
    </div>
  </li>
</DocumentFragment>
`;
